### ISSUE TITLE
[WORK]VisionTransformer

### ISSUE THREAD

**Issue by dustywindow** [Dementia-Diagnosis-Project/memoria_AI#17]
# WORK PLAN 
~ 작업 내용 설명 ~ 
- [X] 모델 : Vision Transformer
  - 선정이유
    a. 
    b. 
  - 사용 프레임워크
  - 데이터
 
- [X] Data Preprocessing

- [X]

### 학습방법
learning rate: 1e-4
batch size : 30
criterion: CrossEntropyLoss
optimizer: Adam

# ISSUES
#### Attention Map Visualization
 Vision Transformer의 Attention Map을 이용하여 학습 결과를 시각화하는 작업

### 예상 해결 방안 list 
- [X] test 이미지를 모델에 맞게 변환하여 input
- [X] attention map 저장
- [ ] attention map들을 하나의 mask로 출력
- [X] 색상변환을 통해 각 픽셀의 영향도를 heatmap 형태로 출력

### (해결시) 해결방법   
- [X] 이미지 한 개만 저장된 test_tensor 구성
 test_loader의 첫 번째 batch의 첫 번째 이미지만 test_tensor에 저장 후 processor로 input 형식에 맞게 preprocessing

- [X] register_forward_hook()을 이용하여 학습된 attention map 저장
  named_parameters()를 이용하여 모델의 각 attention 마지막 레이어 이름 확인 후 해당 레이어 forward_hook
  
- [ ] attention map들을 'mean', 'max', 'mean' 중 하나를 선택해 fusion, discard_ration=0.9 해서 mask 출력
 배경부분에서 시각화부분이 붉게나와 오류가 있는 것 같음. 코드 수정 필요

- [X] 이미지 크기에 맞게 mask를 늘리고 heatmap 형태로 변환
 cv2.resize를 이용해 mask 크기 변환
 cv2.applyColorMap을 이용해 mask를 heatmap형태로 출력
 이미지와 heatmap를 합친 cam 출력

# RESULTS
### Experiment 1
  - A fine-tuned version of [google/vit-base-patch16-224-in21k](https://huggingface.co/google/vit-base-patch16-224-in21k) on Alzheimer MRI data
  - Without Augmentation (only Resize 224*224 and Grayscale to RGB)
 
|Epoch|Step|Training Loss|Validation Loss|Accuracy|Specificity|Sensitivity|AUC|
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|

### Experiment 2
  - A fine-tuned version of [DHEIVER/Alzheimer-MRI](https://huggingface.co/DHEIVER/Alzheimer-MRI) on Alzheimer MRI data
  - Without Augmentation (only Resize 224*224 and Grayscale to RGB)
 
|Epoch|Step|Training Loss|Validation Loss|Accuracy|Specificity|Sensitivity|AUC|
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|1|411|0.7410|0.7115|0.492|0.000|1.000|0.500|
|2|822|0.6885|0.6671|0.596|0.792|0.394|0.593|
|3|1233|0.6475|0.5969|테스트5|테스트6|테스트7|테스트8|
|4|1644|0.5454|0.4787|0.782|0.714|0.851|0.783|
|5|2055|0.4408|0.2934|테스트5|테스트6|테스트7|테스트8|
|6|2466|0.3496|0.3241|테스트5|테스트6|테스트7|테스트8|
|7|2877|0.2831|0.3168|테스트5|테스트6|테스트7|테스트8|
|8|3288|0.2425|0.1755|테스트5|테스트6|테스트7|테스트8|
|9|3699|0.2118|0.0713|테스트5|테스트6|테스트7|테스트8|
|10|4110|0.1835|0.2381|테스트5|테스트6|테스트7|테스트8|
|11|4521|0.1651|0.1966|0.928|0.933|0.922|0.928|
|12|4932|0.1352|0.1561|테스트5|테스트6|테스트7|테스트8|
|13|5343|0.1327|0.2004|테스트5|테스트6|테스트7|테스트8|
|14|5754|0.1182|0.1631|테스트5|테스트6|테스트7|테스트8|
|15|6165|0.1046|0.1480|테스트5|테스트6|테스트7|테스트8|
|16|6576|0.1093|0.1531|테스트5|테스트6|테스트7|테스트8|
|17|6987|0.0977|0.1585|테스트5|테스트6|테스트7|테스트8|
|18|7398|0.0948|0.1340|테스트5|테스트6|테스트7|테스트8|
|19|7809|0.0821|0.1496|테스트5|테스트6|테스트7|테스트8|
|20|8220|0.0831|0.1250|테스트5|테스트6|테스트7|테스트8|
